### ISSUE TITLE
GNU Hurd and LOGIN_NAME_MAX

### ISSUE THREAD

**Issue by alejandro-colomar** [shadow-maint/shadow#1166]
Hi @zeha,

I see that Debian carries a patch for GNU Hurd:
<https://salsa.debian.org/debian/shadow/-/commit/5c6c085ae5ac75cbe8651282f22c6911929a1c66#8970948637e29e415404df854a2e7f8095fe51a2>

which basically does

```diff
+#ifdef __GNU__
+#ifndef LOGIN_NAME_MAX
+#define LOGIN_NAME_MAX 256
+#endif
+#endif
```

I have a couple of questions:

-  Does GNU Hurd not define that limit?
-  What does it return when you call `sysconf(_SC_LOGIN_NAME_MAX);`?

(Depending on the answers to those, I have further questions.)

**Comment by zeha** [comment]
I don't know. IIRC Hurd is against fixed limits.

@sthibaul maybe you can chime in here

**Comment by alejandro-colomar** [comment]
> I don't know. IIRC Hurd is against fixed limits.

That's my suspicion.  And I think that's a bug in Hurd.  At some point you need a limit, even if it is PTRDIFF_MAX.  Claiming no limits is a bug, and can result in Bad Stuff if one ignores the limitations of the system.

> @sthibaul maybe you can chime in here

Thanks!

**Comment by sthibaul** [comment]
Yes, the Hurd usually avoids limits that then end up hard-built in binaries. `sysconf(_SC_LOGIN_NAME_MAX)` returns -1, as posix allows. 

> even if it is PTRDIFF_MAX

A lot of programs which blindly use `FOO_MAX` as allocation size would then just segfault.

> if one ignores the limitations of the system

If they do, they will get ENOMEM and such.

**Comment by alejandro-colomar** [comment]
> Yes, the Hurd usually avoids limits that then end up hard-built in binaries. `sysconf(_SC_LOGIN_NAME_MAX)` returns -1, as posix allows.
> 
> > even if it is PTRDIFF_MAX
> 
> A lot of programs which blindly use `FOO_MAX` as allocation size would then just segfault.
> 
> > if one ignores the limitations of the system
> 
> If they do,

I was meaning that the programmers of the system are ignoring their own limitations.  Not that a programmer might be ignoring them.

The system, being written in the C language, cannot have objects that are bigger than PTRDIFF_MAX bytes.  If the system ignores its own limitations, we're in trouble.

> they will get ENOMEM and such.

I suspect some undefined behavior will trigger somewhere before ENOMEM, and buffer overflows could happen.  I wouldn't want to try.

Why not just specify PTRDIFF_MAX as a hard limit?  It's a hard limit actually, whether you like it or not.  The language doesn't allow more than that safely for any object.

**Comment by sthibaul** [comment]
> Why not just return PTRDIFF_MAX as a hard limit?

> A lot of programs which blindly use FOO_MAX as allocation size would then just segfault.

**Comment by sthibaul** [comment]
> cannot have objects that are bigger than PTRDIFF_MAX bytes. If the system ignores its own limitations, we're in trouble.

Sure, that's for it to take care with sizes. But exposing limitations to programs means hard-building them into binaries.

**Comment by alejandro-colomar** [comment]
> > cannot have objects that are bigger than PTRDIFF_MAX bytes. If the system ignores its own limitations, we're in trouble.
> 
> Sure, that's for it to take care with sizes. But exposing limitations to programs means hard-building them into binaries.

But then, if a user inputs a user name, and we have no length limit (not one that can be checked), the user might input huge strings, and DoS the system.  What's the preferred way to deal with this in Hurd?  Just accept the DoS?  Or maybe add an arbitrary limit in the program?  A third way?

**Comment by sthibaul** [comment]
>  What's the preferred way to deal with this in Hurd?

Limit this within the program and underlying system layers. But not cast it in header interfaces stones.

**Comment by alejandro-colomar** [comment]
> > What's the preferred way to deal with this in Hurd?
> 
> Limit this within the program and underlying system layers. But not cast it in header interfaces stones.

The good thing about having a header is that all programs will share the same limit.  If each program has a different limit, I think the problem is worse than the theoretical problem of not allowing someone to have a 4k username.

**Comment by alejandro-colomar** [comment]
> > > What's the preferred way to deal with this in Hurd?
> > 
> > 
> > Limit this within the program and underlying system layers. But not cast it in header interfaces stones.

Also, I would understand not having the limits cast in stone in headers, but Hurd also doesn't provide limits via sysconf(2), which isn't really in stone.

**Comment by sthibaul** [comment]
> The good thing about having a header is that all programs will share the same limit

The bad thing with such shared limit is that it's then cast in stone. Windows has had MAX_PATH stuck to 260, and escaping that is hell.

> Hurd also doesn't provide limits via sysconf(2),

That depends on the underlying implementation. Some implementations indeed don't actually have limitations.

**Comment by sthibaul** [comment]
> The bad thing with such shared limit [...] escaping that is hell

Notably when libraries expose them in their own ABI.

**Comment by alejandro-colomar** [comment]
> > The good thing about having a header is that all programs will share the same limit
> 
> The bad thing with such shared limit is that it's then cast in stone. Windows has had MAX_PATH stuck to 260, and escaping that is hell.

Makes sense.

> > Hurd also doesn't provide limits via sysconf(2),
> 
> That depends on the underlying implementation. Some implementations indeed don't actually have limitations.

I expect implementations to be limited by PTRDIFF_MAX.  They should probably report that via sysconf(2).  No?
What's the reason to return -1 instead of PTRDIFF_MAX?
Is the system able to go beyond PTRDIFF_MAX?

**Comment by sthibaul** [comment]
> What's the reason to return -1 instead of PTRDIFF_MAX?

Because that's the standard-defined way to say that there is no limitation.

**Comment by alejandro-colomar** [comment]
> > What's the reason to return -1 instead of PTRDIFF_MAX?
> 
> Because that's the standard-defined way to say that there is no limitation.

I would suggest to POSIX that they remove this way of lying to the user, and force systems to reveal their limits, even if they are PTRDIFF_MAX.

**Comment by sthibaul** [comment]
> I would suggest to POSIX that they remove this way of lying to the user, and force systems to reveal their limits, even if they are PTRDIFF_MAX.

I'm not POSIX

**Comment by alejandro-colomar** [comment]
> > I would suggest to POSIX that they remove this way of lying to the user, and force systems to reveal their limits, even if they are PTRDIFF_MAX.
> 
> I'm not POSIX

POSIX doesn't force you to lie (it bogusly allows you to lie).  You can reveal your own limitation of PTRDIFF_MAX.

**Comment by sthibaul** [comment]
> You can reveal your own limitation of PTRDIFF_MAX.

Again, that will just make a lot of programs just either fail to build, or error/segfault out at runtime.

**Comment by alejandro-colomar** [comment]
> > You can reveal your own limitation of PTRDIFF_MAX.
> 
> Again, that will just make a lot of programs just either fail to build, or error/segfault out at runtime.

Can you please develop that?

by returning PTRDIFF_MAX from sysconf(2), I don't see how that could cause a build error.

Regarding the runtime error/segfault, I expect malloc(3) to succeed, and do the real allocation as needed (not allocate the entire PTRDIFF_MAX at once).

It is indeed returning -1 what will cause errors when calling malloc(3).  Unless of course, the program self-limits to a magic number like 42.

**Comment by sthibaul** [comment]
> By returning PTRDIFF_MAX from sysconf(2), I don't see how that could cause a build error.

I was referring to #defining PATH_MAX to PTRDIFF_MAX.

> Regarding the runtime error/segfault, I expect malloc(3) to succeed

?? PTRDIFF_MAX is `2**31` on 32bit systems, and `2**63` on 64bit systems...

**Comment by alejandro-colomar** [comment]
> > By returning PTRDIFF_MAX from sysconf(2), I don't see how that could cause a build error.
> 
> I was referring to #defining PATH_MAX to PTRDIFF_MAX.

Hmm.

> > Regarding the runtime error/segfault, I expect malloc(3) to succeed
> 
> ?? PTRDIFF_MAX is typically `2**31` on 32bit systems, and `2**63` on 64bit systems...

Hmmm.  I thought lazy allocation would make it succeed.  After some experiment, it seems I was wrong.

Okay, so I guess -1 is reasonable then.  I still would prefer a shared limit, but that's just my opinion, of course.  The source code seems to be the same, just that the limit macro needs to be defined per-project, so that's fine.  Thanks!

**Comment by zeha** [comment]
@alejandro-colomar did you end up committing a change to shadow, or do we need to keep carrying the patch?

**Comment by alejandro-colomar** [comment]
> @alejandro-colomar did you end up committing a change to shadow, or do we need to keep carrying the patch?

Hmm, I hadn't.  But I should.  I'll reopen.  Would you mind sending a patch, since it was you who authored it?

**Comment by zeha** [comment]
> I'll send the patch myself.

Thanks

**Comment by alejandro-colomar** [comment]
BTW, CC: @stoeckmann
